src/util/entersState.ts

```diff
@@ -1,59 +1,54 @@
 import { VoiceConnection, VoiceConnectionStatus } from '../VoiceConnection';
 import { AudioPlayer, AudioPlayerStatus } from '../audio/AudioPlayer';
+import { abortAfter } from './abortAfter';
+import EventEmitter, { once } from 'events';
 
 /**
  * Allows a voice connection a specified amount of time to enter a given state, otherwise rejects with an error.
  *
  * @param target - The voice connection that we want to observe the state change for
  * @param status - The status that the voice connection should be in
- * @param maxTime - The maximum time we are allowing for this to occur
+ * @param timeoutOrSignal - The maximum time we are allowing for this to occur, or a signal that will abort the operation
  */
 export function entersState(
 	target: VoiceConnection,
 	status: VoiceConnectionStatus,
-	maxTime: number,
+	timeoutOrSignal: number | AbortSignal,
 ): Promise<VoiceConnection>;
 
 /**
  * Allows an audio player a specified amount of time to enter a given state, otherwise rejects with an error.
  *
  * @param target - The audio player that we want to observe the state change for
  * @param status - The status that the audio player should be in
- * @param maxTime - The maximum time we are allowing for this to occur
+ * @param timeoutOrSignal - The maximum time we are allowing for this to occur, or a signal that will abort the operation
  */
-export function entersState(target: AudioPlayer, status: AudioPlayerStatus, maxTime: number): Promise<AudioPlayer>;
+export function entersState(
+	target: AudioPlayer,
+	status: AudioPlayerStatus,
+	timeoutOrSignal: number | AbortSignal,
+): Promise<AudioPlayer>;
 
 /**
  * Allows a target a specified amount of time to enter a given state, otherwise rejects with an error.
  *
  * @param target - The object that we want to observe the state change for
  * @param status - The status that the target should be in
- * @param maxTime - The maximum time we are allowing for this to occur
+ * @param timeoutOrSignal - The maximum time we are allowing for this to occur, or a signal that will abort the operation
  */
-export function entersState<T extends VoiceConnection | AudioPlayer>(
+export async function entersState<T extends VoiceConnection | AudioPlayer>(
 	target: T,
 	status: VoiceConnectionStatus | AudioPlayerStatus,
-	maxTime: number,
+	timeoutOrSignal: number | AbortSignal,
 ) {
-	if (target.state.status === status) {
-		return Promise.resolve(target);
+	if (target.state.status !== status) {
+		const [ac, signal] =
+			typeof timeoutOrSignal === 'number' ? abortAfter(timeoutOrSignal) : [undefined, timeoutOrSignal];
+		try {
+			await once(target as EventEmitter, status, { signal });
+		} finally {
+			ac?.abort();
```

Comment:
Do we need to call this?

Reply:
It clears the timeout if the state was entered before then